core/src/display_object/video.rs

```diff
@@ -33,56 +33,54 @@ use swf::{DefineVideoStream, VideoCodec, VideoFrame};
 /// framerate.
 #[derive(Clone, Collect, Copy)]
 #[collect(no_drop)]
-pub struct Video<'gc>(GcCell<'gc, VideoData<'gc>>);
+pub struct Video<'gc>(Gc<'gc, VideoData<'gc>>);
 
 impl fmt::Debug for Video<'_> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         f.debug_struct("Video")
-            .field("ptr", &self.0.as_ptr())
+            .field("ptr", &Gc::as_ptr(self.0))
             .finish()
     }
 }
 
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
 pub struct VideoData<'gc> {
-    base: DisplayObjectBase<'gc>,
+    base: RefLock<DisplayObjectBase<'gc>>,
 
-    avm1_text_field_bindings: Vec<Avm1TextFieldBinding<'gc>>,
+    avm1_text_field_bindings: RefLock<Vec<Avm1TextFieldBinding<'gc>>>,
 
     /// The source of the video data (e.g. an external file, a SWF bitstream)
-    source: GcCell<'gc, VideoSource<'gc>>,
+    source: Lock<Gc<'gc, VideoSource<'gc>>>,
 
     /// The decoder stream that this video source is associated to.
-    #[collect(require_static)]
-    stream: VideoStream,
+    stream: Cell<VideoStream>,
 
     /// AVM representation of this video player.
-    object: Option<AvmObject<'gc>>,
+    object: Lock<Option<AvmObject<'gc>>>,
 
     /// List of frames which can be independently seeked to.
     ///
     /// Frames outside of this set must be decoded by playing each frame from
     /// the last keyframe in order. Any out-of-order seeking will be snapped to
     /// the prior keyframe. The first frame in the stream will always be
     /// treated as a keyframe regardless of it being flagged as one.
-    keyframes: BTreeSet<u32>,
+    keyframes: RefCell<BTreeSet<u32>>,
 
     /// The movie whose tagstream or code created the Video object.
     movie: Arc<SwfMovie>,
 
     /// The self bounds for this movie.
-    size: (i32, i32),
+    size: Cell<(i32, i32)>,
 
     /// The last decoded frame in the video stream.
     ///
     /// NOTE: This is only used for SWF-source video streams.
-    #[collect(require_static)]
-    decoded_frame: Option<(u32, BitmapInfo)>,
+    decoded_frame: RefCell<Option<(u32, BitmapInfo)>>,
```

Comment:
This could save a `RefCell` by using a nested struct:
```rs
struct VideoData<'gc> {
  // ...
  cell: RefCell<VideoDataMut>,
}

struct VideoDataMut {
  keyframes: BTreeSet<u32>,
  decoded_frame: Option<(u32, BitmapInfo)>,
}
```

Reply:
I'm not sure I follow. The only reason would be to have one `RefCell` instead of 2? I don't think it has any noticeable perf impact

Reply:
Yup, this is kind of a micro-optimization, but it would save a `RefCell` (and so an `usize` in the struct)

Reply:
That looks like a premature optimization: the memory consumed by the structs >> `usize`, and it's not a free optimization, as we lose the ability to borrow multiple fields independently.